src/graphql/device/Resolvers.js

```diff
@@ -0,0 +1,218 @@
+              type: valueType,
+            });
+          }
+        });
+
+        return (device);
+      } catch (err) {
+        LOG.error(err);
+      }
+    },
+
+    async getDevices(root, params) {
+      // building the request string
+      let requestParameters = {};
+      if (params.hasOwnProperty('page') && params.page.size != null) {
+        requestParameters.page_size = params.page.size;
+      }
+      else {
+        requestParameters.page_size = 20;
+      }
+      if (params.hasOwnProperty('page') && params.page.number != null) {
+        requestParameters.page_num = params.page.number;
+      }
+      else {
+        requestParameters.page_num = 1;
+      }
+      if (params.hasOwnProperty('filter') && params.filter.label != null) {
+        requestParameters.label = params.filter.label;
+      }
+
+      let requestString = '/device?';
+      const keys = Object.keys(requestParameters);
+      const last = keys[keys.length - 1];
+      keys.forEach((element) => {
+        if (element === last) {
+          requestString += `${element}=${requestParameters[element]}`;
+        } else {
+          requestString += `${element}=${requestParameters[element]}&`;
+        }
+      });
+
+      try {
+        const { data: fetchedData } = await axios(optionsAxios(UTIL.GET, requestString));
+        const deviceList = ([{
+          totalCount: fetchedData.devices.length,
+          hasNextPage: fetchedData.pagination.has_next,
+          hasPreviousPage: fetchedData.pagination.page != 1,
+          devices: fetchedData.devices,
+        }]);
+
+        return deviceList;
+
+      } catch (error) {
+        LOG.error(error);
+      }
+    },
+
+    async getDeviceHistory(root, params) {
+      let history = [];
+      const keys = Object.keys(params.filter);
+      keys.shift();
+      const requestStringPt1 = '/history/device/';
+      let requestStringPt2 = '/history';
+
+      if (keys.length != 0) {
+        requestStringPt2 += '?';
+        keys.forEach((element) => {
+          requestStringPt2 += `${element}=${params.filter[element]}&`;
+        });
+      }
+
+      let historyPromiseArray = [];
+      let fetchedData = [];
+      let devicePromiseArray = [];
+      let devicesInfo = [];
+
+      params.filter.devices.forEach((obj) => {
+        obj.attrs.forEach(attr => {
+          let requestString = `${requestStringPt1}${obj.deviceID}${requestStringPt2}&attr=${attr}`;
+          const promiseHistory = axios(optionsAxios(UTIL.GET, requestString)).catch(err => {
+            LOG.error(`Device id ${obj.id}: ${err}`);
+            return Promise.resolve(null);
+          });
+          historyPromiseArray.push(promiseHistory);
+        });
+        const promiseDevice = axios(optionsAxios(UTIL.GET, `/device/${obj.deviceID}`));
+        devicePromiseArray.push(promiseDevice);
+      });
+
+      //API calls are made and results are saved in arrays
+      await Promise.all(historyPromiseArray).then(values => {
```

Comment:
Is everything working fine when are you request more then one and one  of then without history now, when return  just one 404?

Reply:
It is, if there's entries for the other device, it shows only the information about it and ignores entries for the missing one. (The last unit test added is about this scenario)